File: WdtcCore/src/main/java/org/wdt/wdtc/core/auth/UsersList.java
package org.wdt.wdtc.core.auth;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import org.apache.log4j.Logger;
import org.wdt.utils.gson.Json;
import org.wdt.utils.gson.JsonObjectUtils;
import org.wdt.utils.gson.JsonUtils;
import org.wdt.wdtc.core.manger.FileManger;
import org.wdt.wdtc.core.utils.WdtcLogger;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class UsersList {
  private static final File usersListFile = FileManger.getUserListFile();
  private static final Logger logmaker = WdtcLogger.getLogger(UsersList.class);

  public static void addUser(User user) throws IOException {
    JsonObject usersList = getUserListObject();
    String userName = user.getUserName();
    if (usersList.has(userName)) {
      logmaker.warn(userName + " Remove");
      usersList.remove(userName);
    }
    usersList.add(userName, Json.GSON.toJsonTree(user, User.class));
    JsonUtils.writeObjectToFile(usersListFile, usersList, Json.getBuilder().setPrettyPrinting());
  }


  public static JsonObject getUserListObject() throws IOException {
    return JsonUtils.readFileToJsonObject(usersListFile);
  }

  public static User getUser(String UserName) throws IOException {
    return JsonObjectUtils.parseObject(getUserListObject().getAsJsonObject(UserName), User.class);
  }

  public static List<User> getUserList() {
    List<User> userList = new ArrayList<>();
    try {
      Map<String, JsonElement> UserListMap = getUserListObject().asMap();
      if (!UserListMap.keySet().isEmpty()) {
        for (String s : UserListMap.keySet()) {
          userList.add(getUser(s));
        }
      }
    } catch (IOException e) {
      logmaker.error(e);
    }
    return userList;
  }

  public static void printUserList() {
    for (User user : getUserList()) {
      logmaker.info(user);
    }
  }
}

File: WdtcCore/src/main/kotlin/org/wdt/wdtc/core/auth/UsersList.kt
package org.wdt.wdtc.core.auth

import com.google.gson.*
import org.wdt.utils.gson.parseObject
import org.wdt.utils.gson.readFileToClass
import org.wdt.utils.gson.writeObjectToFile
import org.wdt.wdtc.core.auth.UsersList.Companion.saveChangeToFile
import org.wdt.wdtc.core.manger.userJson
import org.wdt.wdtc.core.manger.userListFile
import org.wdt.wdtc.core.utils.gson.TypeAdapters
import org.wdt.wdtc.core.utils.gson.defaultGsonBuilder
import org.wdt.wdtc.core.utils.gson.prettyGsonBuilder
import org.wdt.wdtc.core.utils.info
import org.wdt.wdtc.core.utils.logmaker
import java.lang.reflect.Type


class UsersList(
	private val usersList: LinkedHashSet<User> = linkedSetOf()
) : MutableSet<User> by usersList {
	fun add(loginUser: LoginUser) {
		loginUser.user.let {
			userJson.writeObjectToFile(it, prettyGsonBuilder)
			logmaker.info(it)
			this.add(it)
		}
	}
	
	override fun remove(element: User): Boolean {
		preferredUser.let {
			if (it == element) {
				userJson.writeObjectToFile(JsonObject())
			}
		}
		return usersList.remove(element)
	}
	
	fun LoginUser.addToList() {
		add(this)
	}
	
	companion object {
		fun UsersList.saveChangeToFile() {
			userListFile.writeObjectToFile(this, serializeUsersListGsonBuilder)
		}
	}
}

class UsersListTypeAdapters : TypeAdapters<UsersList> {
	override fun serialize(src: UsersList, typeOfSrc: Type?, context: JsonSerializationContext?): JsonElement {
		return JsonArray().apply {
			src.forEach {
				add(defaultGsonBuilder.create().toJsonTree(it))
			}
		}
	}
	
	override fun deserialize(json: JsonElement, typeOfT: Type?, context: JsonDeserializationContext?): UsersList {
		if (!json.isJsonArray) error("json must be array")
		return LinkedHashSet(json.asJsonArray.map {
			it.asJsonObject.parseObject<User>()
		}).let {
			UsersList(it)
		}
	}
	
}

val serializeUsersListGsonBuilder: GsonBuilder =
	prettyGsonBuilder.registerTypeAdapter(UsersList::class.java, UsersListTypeAdapters())

fun printUserList() = currentUsersList.forEach { logmaker.info(it) }


val currentUsersList: UsersList
	get() = userListFile.readFileToClass(serializeUsersListGsonBuilder)

inline fun UsersList.changeListToFile(block: UsersList.() -> Unit): UsersList {
	return apply(block).also { it.saveChangeToFile() }
}